### PR TITLE
feat(rust): use one socket for all client udp messages

### PR DESCRIPTION
**Pull request by neil2468**
## Current Behavior

The UDP transport currently...
- Uses a new local socket for each UDP message sent to a new destination address.
- Can put the sending part of a socket into an error state, for example, by trying to send to an IPv6 address. Once in the error state, the transport is not able to send more messages. 

## Proposed Changes

Related to #3507. For UDP NAT puncturing it is important that UDP messages sent to an external Rendezvous Service and to an external peer 'B' are all sent from the same UDP address & port. 

Changes to UDP transport...
- Try to use a single local socket for all sent UDP messages which were initiated by an entity within the local node (i.e. so called 'client' messages).
- Enable the senders and listeners to inform the router when they experience an error. On knowing an error has occurred, the router tries to stop and replace the errored sender and listener pair.
- Some refeactoring to reduce code.
- Add tests to exercise the new features.

Apologies for submitting this all in one large chunk. But I couldn't quite work out how not to.

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

Thanks!

### CODE REVIEW

**Comment by davide-baldo** [comment]
Hi! :wave: 
Thanks for this contribution!

**Comment by neil2468** [comment]
Hi 👋 
I just updated the commit message and rebased with develop.
Thanks!

File: implementations/rust/ockam/ockam_transport_udp/src/router/handle.rs [hunk 1]
```diff
@@ -25,99 +11,38 @@ pub(crate) struct UdpRouterHandle {
...
-            hostnames
-                .iter()
-                .filter_map(|s| UdpAddress::from_str(s).ok())
-                .map(|addr| addr.into()),
-        );
-
-        // TODO: should we send a router request instead
-        // and see if worker is already registered?
-        let response = self
+    /// Request reouter start listening on a local UDP port
```
Comment (davide-baldo) [inline implementations/rust/ockam/ockam_transport_udp/src/router/handle.rs]: typo

File: implementations/rust/ockam/ockam_transport_udp/src/router/handle.rs [hunk 2]
```diff
@@ -25,99 +11,38 @@ pub(crate) struct UdpRouterHandle {
...
+    /// so the local node can act as a server to other nodes
+    pub async fn listen(&self, local_addr: SocketAddr) -> Result<()> {
```
Comment (davide-baldo) [inline implementations/rust/ockam/ockam_transport_udp/src/router/handle.rs:34]: nice! :+1:  consistent with TCP side

File: implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs [hunk 3]
```diff
@@ -61,32 +60,41 @@ impl Processor for UdpListenProcessor {
...
                 Ok((msg, addr)) => (msg, addr),
-                Err(_e) => {
-                    info!("Failed to read message from UDP socket.");
+                Err(e) => {
+                    warn!("Failed to read message from UDP socket: {:?}", e);
+
+                    // Inform router of error
+                    let _ = self
+                        .router_handle
+                        .sender_listener_error(self.local_addr)
```
Comment (davide-baldo) [inline implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs]: I'm not sure re-building the UDP socket when it breaks it's the way forward:
> Can put the sending part of a socket into an error state, for example, by trying to send to an IPv6 address.

Can't we avoid breaking the socket by allowing only ipv4? Are there other cases?

File: implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs [hunk 4]
```diff
@@ -61,32 +60,41 @@ impl Processor for UdpListenProcessor {
...
-        self.router_handle
-            .register(self.tx_addr.clone(), addr.to_string())
-            .await?;
-
-        msg.return_route.modify().prepend(UdpAddress::from(addr));
-
-        debug!("Message onward route: {}", msg.onward_route);
-        debug!("Message return route: {}", msg.return_route);
+        // Set return route to go directly to paired sender, skipping the UDP router
+        msg.return_route
```
Comment (davide-baldo) [inline implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs]: You can use the `route!` macro to make it a bit more readable:
```rust
msg.return_route = route![
  self.sender_addr.clone(),
  Address::new(UDP, addr.to_string()),
  msg.return_route
];
```

File: implementations/rust/ockam/ockam_transport_udp/src/workers/sender.rs [hunk 5]
```diff
@@ -38,24 +43,54 @@ impl Worker for UdpSendWorker {
...
+                trace!("Successful send to {}", addr);
+                Ok(())
+            }
+            Err(e) => {
+                warn!("Failed send to {}: {:?}", addr, e);
+
+                // Inform router of our error
+                let _ = self
+                    .router_handle
+                    .sender_listener_error(self.local_addr)
```
Comment (davide-baldo) [inline implementations/rust/ockam/ockam_transport_udp/src/workers/sender.rs]: Same as before, can we avoid breaking it?

File: implementations/rust/ockam/ockam_transport_udp/tests/tests.rs [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+use ockam_core::{route, AllowAll, Result, Routed, Worker};
+use ockam_node::Context;
+use ockam_transport_udp::{UdpTransport, UDP};
+use std::net::SocketAddr;
+use std::time::Duration;
+use tokio::net::UdpSocket;
+use tracing::{debug, error, trace};
+
+const TIMEOUT: Duration = Duration::from_secs(5);
+const AVAILABLE_LOCAL_PORTS_ADDR: &str = "0.0.0.0:0";
```
Comment (davide-baldo) [inline implementations/rust/ockam/ockam_transport_udp/tests/tests.rs]: Can you bind to 127.0.0.1 rather than 0.0.0.0?

**Review by davide-baldo — CHANGES_REQUESTED** [review]
Nice step up! :+1:

**Comment by davide-baldo** [comment]
@neil2468 
I wanted to let you know we are planning to remove auto-connect, and protocol from `Address` to avoid security pitfalls.
So, the router will change heavily yet again, and you'll need to explicitly create outgoing workers as you need to declare listeners.
Unfortunately, I can't give you any fixed time frame yet.

**Comment by mrinalwadhwa** [comment]
> @neil2468
I wanted to let you know we are planning to remove auto-connect, and protocol from Address to avoid security pitfalls.
So, the router will change heavily yet again, and you'll need to explicitly create outgoing workers as you need to declare listeners.
Unfortunately, I can't give you any fixed time frame yet.

@neil2468 I think we can move these two efforts in parallel. It would be best for you to keep building on the current code and for us to start landing the changes. We can refactor for the design changes @davide-baldo mentioned above as we go.

File: implementations/rust/ockam/ockam_transport_udp/src/router/handle.rs (hunk 2 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_transport_udp/src/router/handle.rs:34]: That was lucky 😄 I hadn't noticed.

File: implementations/rust/ockam/ockam_transport_udp/src/router/handle.rs (hunk 1 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_transport_udp/src/router/handle.rs]: Done, thanks.

File: implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs (hunk 4 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs]: Done.

File: implementations/rust/ockam/ockam_transport_udp/tests/tests.rs (hunk 6 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_transport_udp/tests/tests.rs]: Okay.

File: implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs (hunk 3 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs]: I've seen the sink part of a socket get in to an error state when (1) sending to an IPv6 address and (2) when sending to "0.0.0.0:0". The transport does not support IPv6, so (1) can no longer happen, but something like (2) or something I've never seen could happen.

My gut feeling is that errors will be rare as UDP is happy to send to peers that have no route or are disconnected. However, I don't know what errors could happen on all platforms and now, as the transport uses a single UDP socket for 'client' messages to all peers, if an error occurs the transport will grind to a halt for 'client' messages unless the transport tries to clear/recover from the error.

I'm happy to implement whatever Ockam prefer. 👍

File: implementations/rust/ockam/ockam_transport_udp/src/workers/sender.rs (hunk 5 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_transport_udp/src/workers/sender.rs]: See related comments on `listener.rs`.

File: implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs (hunk 3 above)
Comment (davide-baldo) [reply implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs]: > as the transport uses a single UDP socket for 'client' messages to all peers, if an error occurs the transport will grind to a halt for 'client' messages unless the transport tries to clear/recover from the error.

I 100% agree, we cannot afford to have a broken socket when the whole node depends on it.
Still, what worries me most, is the scenario where one worker keeps re-trying sending a message and keeps breaking the socket. In this case, the service seems to be working fine and will probably go undetected, but the packet drop is higher than it should be.

> However, I don't know what errors could happen on all platforms

How about fixing errors one step at the time? I believe a broken socket doesn't go unnoticed, and will help identify and fix every failure case in every platform.

If you could please reject and log messages sent toward `0.0.0.0`, we could omit the recover mechanism.

Comment (neil2468) [reply implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs]: Okay, cool 👍 .

I've added the below check in `UdpSendWorker` just before sending, and have removed the recover mechanism.

```
        // Error on conditions that _might_ put the sink
        // into an error state
        if addr.port() == 0 {
            warn!(peer_addr = %peer_addr, "Will not send to address");
            return Err(TransportError::InvalidAddress.into());
        }
```

From testing, it seems sending to a port 0 causes the sink to get into an error state. For example, sending to `0.0.0.0:8080` is okay but `0.0.0.0:0` and `192.168.1.10:0` are not.

**Review by davide-baldo — APPROVED** [review]
Nice! You even anticipated the auto-connect removal and simplified the architecture by an order of magnitude, hat off! :+1:

**Pull request merged**